src/sympc/tensor/grads/grad_functions.py

```diff
@@ -432,6 +435,159 @@ def backward(ctx: Dict[str, Any], grad: MPCTensor) -> MPCTensor:
+            grad_output=grad_output.tensor,
+            input_size=input_size,
+            stride=(stride, stride),
+            padding=(padding, padding),
+            kernel_size=kernel_size,
+            dilation=(dilation, dilation),
+        )
+
+        share_tensor = ShareTensor(torch.tensor(new_tuple), session=session)
+        return share_tensor
+
+    @staticmethod
+    def forward(
+        ctx: Dict[str, Any],
+        input: Union["MPCTensor", torch.Tensor, float, int],
+        weight: Union["MPCTensor", torch.Tensor, float, int],
+        bias: None,
+        stride: int,
+        padding: int,
+        dilation: int,
+        groups: int,
+    ) -> MPCTensor:
+        """Perform the feedforward and compute the result for the conv2d operation.
+
+        Args:
+            ctx (Dict[str, Any]): Context used to save information needed in the backward pass
+            input: the input
+            weight: the convolution kernel
+            bias: optional bias
+            stride: stride
+            padding: padding
+            dilation: dilation
+            groups: groups
+
+        Returns:
+            (MPCTensor): The result of the conv2d operation
+        """
+        ctx["input"] = input
+        ctx["weight"] = weight
+        ctx["stride"] = stride
+        ctx["padding"] = padding
+        ctx["dilation"] = dilation
+        ctx["groups"] = groups
+
+        return input.conv2d(weight, bias, stride, padding, dilation, groups)
+
+    @staticmethod
+    def backward(ctx: Dict[str, Any], grad: MPCTensor) -> Tuple[MPCTensor]:
+        """Perform the backward pass for the conv2d operation.
+
+        Args:
+            ctx (Dict[str, Any]): Context used to retrieve the information for the backward pass
+            grad (MPCTensor): The gradient that came from the child nodes
+
+        Returns:
+            (input_grad, weight_grad) (Tuple[MPCTensor]): The gradients passed
+            to the input and kernal nodes.
+        """
+        input = ctx["input"]
+        weight = ctx["weight"]
+        stride = ctx["stride"]
+        padding = ctx["padding"]
+        dilation = ctx["dilation"]
+        groups = ctx["groups"]
+        weight_size = (weight.shape[2], weight.shape[3])
+        in_channels = input.shape[1]
+        out_channels = grad.shape[1]
+        min_batch = input.shape[0]
+
+        common_args = [
+            tuple(input.shape),
+            stride,
+            padding,
+            weight_size,
+            dilation,
+            grad.session,
+        ]
+        args = [[el] + common_args for el in grad.share_ptrs]
+
+        shares = parallel_execution(GradConv2d.function_aux, grad.session.parties)(args)
+        grad_input_padding = MPCTensor(shares=shares, session=grad.session)
+
+        output_padding_tensor = grad_input_padding.reconstruct()
+        output_padding = tuple(output_padding_tensor.to(torch.int).tolist())
+
+        input_grad = grad.conv_transpose2d(
```

Comment:
Q: Does this?

Reply:
No if the parties > 2 we get the over-scaled value we talked about.